File: QuantConnect/Strategy_17.py


# Trend Following Strategy using Donchian Channels, with Entry on Price Breakout, Exit on Opposite Signal or Trailing Stop, and Stop Loss of 1 x ATR.

from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import DonchianChannel
from datetime import timedelta, datetime
import decimal as d

class TrendFollowingStrategy(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2010, 1, 1)
        self.SetEndDate(2020, 12, 31)
        self.SetCash(100000)
        self.symbol = "SPY"
        self.period = 20
        self.atr_period = 14
        self.atr_multiplier = 1
        self.donchian_upper = None
        self.donchian_lower = None
        self.entry_price = None
        self.exit_price = None
        self.stop_loss_price = None
        
        # Set Donchian Channels as Primary Indicator for Strategy Entry and Exit Signals.
        
        self.donchian_channel = DonchianChannel(self.period)
        
        # Set Stop Loss on ATR at Entry Price.
        
        self.atr_indicator = AverageTrueRange(self.atr_period)
        
    def OnData(self, data):
        
        # Wait until we have enough data for our indicators.
        
        if not (self.donchian_channel.IsReady and 
                self.atr_indicator.IsReady): 
            return
        
        # Define our Entry Criteria as Price Breaking Out of Upper Donchian Channel.
        
            if data[self.symbol].Close > self.donchian_channel.UpperBand.Current.Value:
                
                # Set Entry Price at Upper Band Value.
                
                self.entry_price = data[self.symbol].Close
                
                # Set Stop Loss Price at Entry Price - ATR Multiplier x ATR Value.
                
                atr_value = float(self.atr_indicator.Current.Value)
                stop_loss_offset = atr_value * float(self.atr_multiplier)
                stop_loss_decimal_places = len(str(int(atr_value)))
                stop_loss_decimal_places += 2
                
                stop_loss_price_decimal_places_shifted_right = d.Decimal(stop_loss_offset).shift(stop_loss_decimal_places) 
                stop_loss_price_decimal_places_shifted_left_as_string_formatting_two_decimals_and_adding_negative_sign_to_left_shifted_value \
                    = "-{0:.2f}".format(stop_loss_price_decimal_places_shifted_right) 
                
                stop_loss_price_string_shifted_left_with_two_decimals_and_negative_sign_as_first_character \
                    = "stop loss: " + str(stop_loss_price_decimal_places_shifted_left_as_string_formatting_two_decimals_and_adding_negative_sign_to_left_shifted_value) 
                
                log_message_stop_loss_set_at_this_stop_level \
                    = "stop loss set at: " + str(float(self.entry_price) - float(stop_loss_offset))
                    
                # Debugging Print Statements.
                    
#                 print(log_message_stop_loss_set_at_this_stop_level)
#                 print(stop_loss_price_string_shifted_left_with_two_decimals_and_negative_sign_as_first_character)

                    
            # Define our Exit Criteria as Opposite Signal or Trailing Stop Loss Hit.

            if (data[self.symbol].Close < self.donchian_channel.LowerBand.Current.Value or 
               data[self.symbol].Close > (float(self.entry_price) + (float(atr_value) * float(self.atr_multiplier)))):
                
                # Reset All Variables to Begin New Trade Cycle.
                
                    entry_exit_log_message_what_was_the_entry_strategy \
                        ="entry strategy: donchain channel breakout"
                        
                    entry_exit_log_message_what_was_the_exit_strategy \
                        ="exit strategy: opposite signal or trailing stop"
                        
                    entry_exit_log_message_entry_trade_time \
                        ="entry time:" + str(data[self.symbol].Time.date())
                        
                    entry_exit_log_message_entry_trade_opening_condition \
                        ="entry trade opening condition: price crossing upper band value."
                        
                    entry_exit_log_message_entry_trade_opening_condition_additional_info \
                        ="Upper band value: " + str(float(self.donchian_channel.UpperBand.Current.Value))
                        
                    entry_exit_log_message_exit_trade_time \
                        ="exit time:" + str(data[self.symbol].Time.date())
                        
                    if data[self.symbol].Close < float(self.donchian_channel.LowerBand.Current.Value):
                        
                        exit_signal_type_that_triggered_the_exit_condition \
                            ="exit signal type that triggered the exit condition:" + "price crossing lower band value."
                            
                        exit_signal_type_that_triggered_the_exit_condition_additional_info\
                            ="lower band value crossed:" +" "+ str(float(self.donchian_channel.LowerBand.Current.Value))
                            
                            
                            
                    elif data[self.symbol].Close > (float(self.entry_price) + (float(atr_value) * float(self.atr_multiplier))):
                        
                        exit_signal_type_that_triggered

File: QuantConnect/Strategy_17.cs


# Strategy_TrendFollowing.py

from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import DonchianChannel
from datetime import timedelta

class TrendFollowing(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2010, 1, 1)
        self.SetEndDate(2020, 12, 31)
        self.SetCash(100000)
        self.AddEquity("SPY", Resolution.Daily)

        # Initialize Donchian Channel
        self.dc = DonchianChannel(20)

        # Initialize ATR
        self.atr = self.ATR("SPY", 20, MovingAverageType.Simple, Resolution.Daily)

        # Initialize last trade time and price
        self.lastTradeTime = None
        self.lastTradePrice = None

    def OnData(self, data: TradeBar):
        
        # Wait for Donchian Channel to fully initialize
        if not self.dc.IsReady:
            return

        if not self.Portfolio.Invested:
            if data.Close > self.dc.UpperBand.Current.Value:
                # Buy when price breaks above upper band of Donchian Channel
                self.SetHoldings("SPY", 1.0)
                self.lastTradeTime = data.Time
                self.lastTradePrice = data.Close

            elif data.Close < self.dc.LowerBand.Current.Value:
                # Short when price breaks below lower band of Donchian Channel
                self.SetHoldings("SPY", -1.0)
                self.lastTradeTime = data.Time
                self.lastTradePrice = data.Close
        
        else:
            # Check for exit criteria
            
            if (self.Portfolio.Invested > 0 and data.Close < self.dc.LowerBand.Current.Value) or (self.Portfolio.Invested < 0 and data.Close >self.dc.UpperBand.Current.Value):
                # Opposite signal received, exit position 
                if (self.Portfolio.Invested > 0):
                    # Long position to exit 
                    stopPrice = max(self.lastTradePrice - (self.atr.Current.Value),data.Low)
                    orderTicket=self.StopMarketOrder("SPY",-self.Portfolio["SPY"].Quantity,stopPrice)
                    if orderTicket is not None and orderTicket.Status==OrderStatus.Filled:
                        Log(f"Exit long position at {orderTicket.AverageFillPrice} on {data.Time}")
                        return 
                    
                
                elif (self.Portfolio.Invested < 0):
                    # Short position to exit 
                    stopPrice = min(self.lastTradePrice + (self.atr.Current.Value),data.High) 
                    orderTicket=self.StopMarketOrder("SPY",-self.Portfolio["SPY"].Quantity,stopPrice) 
                    
                    if orderTicket is not None and orderTicket.Status==OrderStatus.Filled:
                        Log(f"Exit short position at {orderTicket.AverageFillPrice} on {data.Time}")
                        return 

            else:
                
                 ## Trailing Stop Loss - Long Position
                
                
                 if (self.Portfolio.Invested > 0):   
                     
                     stopLoss= max(self.lastTradePrice - (1 *self.atr.Current.Value),data.Low)

                     ## Check trailing stop loss condition                 
                     if(data.Low < stopLoss):               
                     
                         orderTicket=self.StopMarketOrder("SPY",-self.Portfolio["SPY"].Quantity,stopLoss) 
                         
                         if orderTicket is not None and orderTicket.Status==OrderStatus.Filled:
                            Log(f"Stop Loss Triggerred for long position at {orderTicket.AverageFillPrice} on {data.Time}")
                            return 

                 ## Trailing Stop Loss - Short Position
                
                 elif (self.Portfolio.Invested < 0):   
                     
                     stopLoss= min(self.lastTradePrice + (1 *self.atr.Current.Value),data.High)

                     ## Check trailing stop loss condition                 
                     if(data.High > stopLoss):               
                     
                         orderTicket=self.StopMarketOrder("SPY",-self.Portfolio["SPY"].Quantity,stopLoss) 
                         
                         if orderTicket is not None and orderTicket.Status==OrderStatus.Filled:
                            Log(f"Stop Loss Triggerred for short position at {orderTicket.AverageFillPrice} on {data.Time}")
                            return 

        
    def OnEndOfDay(self):
        
         ## Update the last trade price when there are no trades during the day.
         
         holdings=self.Securities["SPY"].Holdings
        
         if holdings is not None and holdings.Quantity!=0:            
             ## Update the last trade time when there are no trades during the day.
             lastbar=self.History(["SPY"],2).iloc[0]             
             symbol="SPY"
             last_price=lastbar.loc[symbol]['close']           
             
             ## Update the